sync2/rangesync/message.go

```diff
@@ -0,0 +1,210 @@
+package rangesync
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/sync2/types"
+)
+
+// MessageType specifies the type of a sync message.
+type MessageType byte
+
+const (
```

Comment:
> the set of protocol messages is huge for a p2p protocol

I will think about reducing the number of messages. But it can't probably be reduced "too much".
There are three groups of messages:
1. Set reconciliation
2. Probe (minhash)
3. Recent pre-sync (priming the sync to lower the difference by sending recent messages)
I will try to make that part more clear

> their sequence is not documented in the code
There's no real sequence right now, as `RangeSetReconciler` has no state. Each message receives a response based on its type. There's something to be said about which messages are used to initiate synchronization, though.
I will add separate protocol docs as `set-reconciliation.md` to clear things up

> there's not even a diagram about the relationship between the types of messages in the research paper which is linked

The doc itself is not very clear on message types indeed. Also it doesn't include the minhash and recent sync part which is not part of the recursive set reconiliation algorithm. I will try describe this (and add some diagrams) in `set-reconciliation.md`

> the intersecting point of the above, which is potentially a moot point, but in my opinion - usually generates attack surfaces. personally i find complex p2p protocols to be very difficult to reason about, and very difficult to maintain and onboard onto.

Well, I'd be quite glad to simplify it, and I'm open to suggestions how to do it.
Things that come to mind
1) Remove minhash: errors in gauging the differences between peers by using mere item counts may cause performance problems. Also, minhash per se is not really that complex
2) Remove recent sync: this may have rather severe negative performance impact during cycle gaps
3) Simplify the set reconciliation itself: not sure it can get much simpler, frankly
4) Do not do set reconciliation at all: we're stuck with existing sync which is quite resource hungry, even if we improve it somewhat to use multipeer split sync

What I'd add is fuzz tests. Probably I need to add it to the list of PRs we must complete before using `syncv2` on the mainnet.

Reply:
so about simplifying: there are some messages, like empty set or empty range that can simply be represented by an explicit state over the protocol messages that represent a response to whether there are items in the set or range (i.e. repurpose existing messages to convey the message)

Reply:
also about fuzzing the protocol: I think this is going to require a lot of effort to write, and may require a massive effort to later understand and fix (in case tests break). But that's just me

Reply:
Docs added, including message type descriptions: https://github.com/spacemeshos/go-spacemesh/blob/syncv2/pairwise/dev-docs/sync2-set-reconciliation.md

Re `EmptySet` and `EmptyRange`: `EmptyRange` is basically `RangeContents` with zero count, but due to message type field being needed anyway, by using it to specify that the range is empty we save a byte needed for the zero count otherwise, and while this is not a huge difference, it still contributes to the amount of data sent during sync, as in some cases there can be a lot of `EmptyRange` messages for small ranges.

`EmptySet` could be made a special case of `EmptyRange` with zero-length `X` and `Y` fields. I'll further check if it causes additional difficulties or not.